### ISSUE TITLE
[PATCH]twisted.web.static.addSlash should preserve the query string

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7593]
|<img alt="jolby's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jolby reported|
|-|-|
|Trac ID|trac#364|
|Type|defect|
|Created|2003-10-22 10:46:33Z|

Attachments:

* [static.addSlash.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/56e/56e43ae4ca9369ef504ed49d4a92f42eddff81c5/583bfd3cbd19f703a9262d8ed17da64ffc6852a8.patch) (600 bytes) - added by jolby on 2003-10-22 10:46:33Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__364 364
type__defect defect
reporter__jolby jolby
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1066819593000000 1066819593000000
changetime__1469007858388463 1469007858388463
version__ 
owner__ 
cc__radix cc__jolby
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jolby's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jolby commented|
|-|-|

```
#!html
<pre>
The addSlash of twisted.web.static.addSlash strips out the query string of a
request. This patch should fix the problem and retain the query string.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
please spell "patch" in all-caps for easier searching
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Assigning this to someone... patches shouldn't be forgotten for this long!
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
I've known this bug was around, but it doesn't interest me. Why not just send
GET to the correct URL?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="jolby's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jolby commented|
|-|-|

```
#!html
<pre>
I'm not really sure what you mean by "just send GET to the correct URL", so I'll
just state my case as to why I think this patch should go in:
1. This breaks login behavior because guard.SessionWrapper uses addSlash to
redirect to the originally requested resource after login. Let's say you have a
resource path /blah/viewuser. The viewuser resource expects a request arg:
id=&lt;someid>. If you have your site resource wrapped in a guard.SessionWrapper,
it won't render correctly after login is complete. If you have a user who has
their session time out, and they try to go to /blah/viewuser/?id=2, they will be
redirected to /blah/viewuser/ after logging in. With this patch they will be
correctly redirected to /blah/viewuser?id=2
2. The current behavior is surprising I think. The name of the method suggests
that it will just return the passed-in request uri with a slash appended to the
last resource, when it also strips out the query string part of the uri. If this
patch isn't accepted, it might be clearer to rename addSlash to something more
accurate like, addSlashAndStripQueryString.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="jolby's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jolby commented|
|-|-|

```
#!html
<pre>
Errr-- on my previous post, the last sentence of the 2nd paragraph should read:

With this patch they will be correctly redirected to /blah/viewuser/?id=2

(path has slash appended)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Ok. Sorry for leaving this sit around again.

Your use case was convincing, and I don't *think* the patch will break anything,
so I've committed it.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|GitHub commented|
|-|-|

In changeset 14fd13a32e741a5bc443dbb662dc405306609319
```
#!CommitTicketReference repository="" revision="14fd13a32e741a5bc443dbb662dc405306609319"
Merge pull request #364 from twisted/8630-scop-spelling

Author: scop
Reviewer: rodrigc
Fixes: #8630

Spelling fixes
```
